---
title: Effectの割り込みモデルの紹介
navTitle: Interruption Model
excerpt: Effectの割り込みモデルの複雑さを探求し、並行アプリケーション開発における重要な側面を学びます。親ファイバーが子ファイバーを終了させるシナリオ、競争するファイバー、ユーザーによる割り込み、タイムアウトなど、ファイバーの割り込み処理のニュアンスを理解します。ポーリングと非同期割り込みの比較を掘り下げ、後者の一貫性維持と機能的パラダイムの遵守における利点を理解します。ファイバーが割り込まれるタイミングについての洞察を得て、この重要な機能の包括的な理解を提供する例とシナリオを示します。
bottomNavigation: pagination
---

## ファイバーの割り込み処理

並行アプリケーションを開発する際には、他のファイバーの実行を割り込む必要がある場合がいくつかあります。例えば：

1. 親ファイバーがタスクを実行するためにいくつかの子ファイバーを開始し、後で親ファイバーが一部またはすべての子ファイバーの結果を必要としないと決定する場合。

2. 二つ以上のファイバーが互いに競争を開始する場合。最初に結果が計算されたファイバーが勝ち、他のすべてのファイバーは不要となり、割り込まれるべきです。

3. インタラクティブなアプリケーションでは、ユーザーが既に実行中のタスクを停止したい場合があります。例えば、「停止」ボタンをクリックしてファイルのダウンロードを防ぐ場合などです。

4. 予想よりも長く実行される計算は、タイムアウト操作を使用して中止する必要があります。

5. ユーザー入力に基づいて計算集約型タスクを実行するアプリケーションでは、ユーザーが入力を変更した場合、現在のタスクをキャンセルして別のタスクを実行する必要があります。

## ポーリング vs. 非同期割り込み

ファイバーを割り込む際、単純なアプローチは一つのファイバーが他のファイバーを強制的に終了させることです。しかし、このアプローチは、ターゲットファイバーが共有状態を変更している途中であれば、共有状態が一貫性のない信頼できない状態に陥る可能性があるため理想的ではありません。したがって、共有の可変状態の内部一貫性を保証しません。

代わりに、この問題に対処するための二つの一般的で有効な解決策があります：

1. **半非同期割り込み（割り込みのためのポーリング）**：命令型言語は、Javaのように、半非同期のシグナリングメカニズムとしてポーリングを使用することがよくあります。このモデルでは、ファイバーが他のファイバーに割り込み要求を送信します。ターゲットファイバーは割り込みステータスを継続的にポーリングし、他のファイバーから割り込み要求を受信したかどうかを確認します。割り込み要求が検出されると、ターゲットファイバーはできるだけ早く自分自身を終了します。

   この解決策では、ファイバー自体がクリティカルセクションを処理します。したがって、ファイバーがクリティカルセクションの途中で割り込み要求を受け取った場合、割り込みを無視し、クリティカルセクションが終了するまでその処理を延期します。

   しかし、このアプローチの欠点は、プログラマーが定期的にポーリングするのを忘れると、ターゲットファイバーが応答しなくなり、デッドロックを引き起こす可能性があることです。さらに、グローバルフラグのポーリングはEffectが従う機能的パラダイムと一致しません。

2. **非同期割り込み**：非同期割り込みでは、ファイバーが他のファイバーを終了させることが許可されます。ターゲットファイバーは割り込みステータスをポーリングする責任を負いません。代わりに、クリティカルセクション中にターゲットファイバーはその領域の割り込み可能性を無効にします。これはグローバル状態のポーリングを必要としない純粋に機能的な解決策です。Effectはその割り込みモデルにこの解決策を採用しており、完全に非同期のシグナリングメカニズムです。

   このメカニズムは、定期的にポーリングするのを忘れるという欠点を克服します。また、純粋に機能的な計算では、クリティカルセクション中に割り込みが無効にされている場合を除き、任意の時点で計算を中止することができるため、機能的パラダイムと完全に互換性があります。

## ファイバーが割り込まれるタイミング

ファイバーが割り込まれる方法や状況はいくつかあります。それぞれを探り、これらのシナリオを再現する方法を示す例を提供します。

### Effect.interruptの呼び出し

特定のファイバーに対して`Effect.interrupt`オペレーターを呼び出すことで、ファイバーを割り込むことができます。

割り込みなし

<Tabs items={["Effect.genを使用", "pipeを使用"]}>
<Tab>

```ts twoslash
import { Effect } from "effect"

const program = Effect.gen(function* () {
  yield* Effect.log("start")
  yield* Effect.sleep("2 seconds")
  yield* Effect.log("done")
})

Effect.runPromise(program).catch((error) =>
  console.log(`interrupted: ${error}`)
)
/*
出力:
timestamp=... level=INFO fiber=#0 message=start
timestamp=... level=INFO fiber=#0 message=done
*/
```

</Tab>
<Tab>

```ts twoslash
import { Effect } from "effect"

const program = Effect.log("start").pipe(
  Effect.andThen(Effect.sleep("2 seconds")),
  Effect.andThen(Effect.log("done"))
)

Effect.runPromise(program).catch((error) =>
  console.log(`interrupted: ${error}`)
)
/*
出力:
timestamp=... level=INFO fiber=#0 message=start
timestamp=... level=INFO fiber=#0 message=done
*/
```

</Tab>
</Tabs>

割り込みあり

<Tabs items={["Effect.genを使用", "pipeを使用"]}>
<Tab>

```ts {6} twoslash
import { Effect } from "effect"

const program = Effect.gen(function* () {
  yield* Effect.log("start")
  yield* Effect.sleep("2 seconds")
  yield* Effect.interrupt
  yield* Effect.log("done")
})

Effect.runPromiseExit(program).then(console.log)
/*
出力:
timestamp=... level=INFO fiber=#0 message=start
{
  _id: 'Exit',
  _tag: 'Failure',
  cause: {
    _id: 'Cause',
    _tag: 'Interrupt',
    fiberId: {
      _id: 'FiberId',
      _tag: 'Runtime',
      id: 0,
      startTimeMillis: ...
    }
  }
}
*/
```

</Tab>
<Tab>

```ts {5} twoslash
import { Effect } from "effect"

const program = Effect.log("start").pipe(
  Effect.andThen(Effect.sleep("2 seconds")),
  Effect.andThen(Effect.interrupt),
  Effect.andThen(Effect.log("done"))
)

Effect.runPromiseExit(program).then(console.log)
/*
出力:
timestamp=... level=INFO fiber=#0 message=start
{
  _id: 'Exit',
  _tag: 'Failure',
  cause: {
    _id: 'Cause',
    _tag: 'Interrupt',
    fiberId: {
      _id: 'FiberId',
      _tag: 'Runtime',
      id: 0,
      startTimeMillis: ...
    }
  }
}
*/
```

</Tab>
</Tabs>

### 複数のエフェクトの割り込み

`Effect.forEach`のような関数を使って複数のエフェクトを組み合わせるとき、一つのエフェクトが割り込まれると、他のすべてのエフェクトも割り込まれることに注意が必要です。例を見てみましょう：

<Tabs items={["Effect.genを使う", "pipeを使う"]}>
<Tab>

```ts twoslash
import { Effect } from "effect"

const program = Effect.forEach(
  [1, 2, 3],
  (n) =>
    Effect.gen(function* () {
      yield* Effect.log(`start #${n}`)
      yield* Effect.sleep(`${n} seconds`)
      if (n > 1) {
        yield* Effect.interrupt
      }
      yield* Effect.log(`done #${n}`)
    }),
  { concurrency: "unbounded" }
)

Effect.runPromiseExit(program).then((exit) =>
  console.log(JSON.stringify(exit, null, 2))
)
/*
出力:
timestamp=... level=INFO fiber=#1 message="start #1"
timestamp=... level=INFO fiber=#2 message="start #2"
timestamp=... level=INFO fiber=#3 message="start #3"
timestamp=... level=INFO fiber=#1 message="done #1"
{
  "_id": "Exit",
  "_tag": "Failure",
  "cause": {
    "_id": "Cause",
    "_tag": "Parallel",
    "left": {
      "_id": "Cause",
      "_tag": "Interrupt",
      "fiberId": {
        "_id": "FiberId",
        "_tag": "Runtime",
        "id": 3,
        "startTimeMillis": 1712582275153
      }
    },
    "right": {
      "_id": "Cause",
      "_tag": "Sequential",
      "left": {
        "_id": "Cause",
        "_tag": "Empty"
      },
      "right": {
        "_id": "Cause",
        "_tag": "Interrupt",
        "fiberId": {
          "_id": "FiberId",
          "_tag": "Runtime",
          "id": 0,
          "startTimeMillis": 1712582275147
        }
      }
    }
  }
}
*/
```

</Tab>
<Tab>

```ts twoslash
import { Effect } from "effect"

const program = Effect.forEach(
  [1, 2, 3],
  (n) =>
    Effect.log(`start #${n}`).pipe(
      Effect.andThen(() => {
        const effect = Effect.sleep(`${n} seconds`)
        if (n > 1) {
          return Effect.andThen(effect, () => Effect.interrupt)
        } else {
          return effect
        }
      }),
      Effect.andThen(Effect.log(`done #${n}`))
    ),
  { concurrency: "unbounded" }
)

Effect.runPromiseExit(program).then((exit) =>
  console.log(JSON.stringify(exit, null, 2))
)
/*
出力:
timestamp=... level=INFO fiber=#1 message="start #1"
timestamp=... level=INFO fiber=#2 message="start #2"
timestamp=... level=INFO fiber=#3 message="start #3"
timestamp=... level=INFO fiber=#1 message="done #1"
{
  "_id": "Exit",
  "_tag": "Failure",
  "cause": {
    "_id": "Cause",
    "_tag": "Parallel",
    "left": {
      "_id": "Cause",
      "_tag": "Interrupt",
      "fiberId": {
        "_id": "FiberId",
        "_tag": "Runtime",
        "id": 3,
        "startTimeMillis": 1712582275153
      }
    },
    "right": {
      "_id": "Cause",
      "_tag": "Sequential",
      "left": {
        "_id": "Cause",
        "_tag": "Empty"
      },
      "right": {
        "_id": "Cause",
        "_tag": "Interrupt",
        "fiberId": {
          "_id": "FiberId",
          "_tag": "Runtime",
          "id": 0,
          "startTimeMillis": 1712582275147
        }
      }
    }
  }
}
*/
```

</Tab>
</Tabs>

この例では、3つの並行タスクを表す配列 `[1, 2, 3]` があります。`Effect.forEach` を使用して各要素を反復処理し、いくつかの操作を実行します。`Effect.log` 関数は、各タスクの開始と完了を示すメッセージをログに記録するために使用されます。

出力を見ると、`n = 1` のタスクは正常に開始して完了することがわかります。しかし、`n = 2` のタスクは完了する前に `Effect.interrupt` を使用して中断されます。その結果、すべてのファイバーが中断され、プログラムは "All fibers interrupted without errors" というメッセージとともに終了します。

この例は、並行エフェクトでの中断がどのように機能するかを示しています。並行タスクの1つが中断されると、他のすべての並行タスクの中断を引き起こします。
